Universal Links - Add navigation to an article selected from a widget

Gio2018 opened the pull request:
## Summary
* This PR lets you tap on an article in a widget and show it in the app.

## References 
* Branch name

## Implementation Details
* Add `UrlValidator`, a type that validates urls against the supported universal links format(s)
* Update `MainViewModel`, add `handle(_ url: URL)`, which takes urls intercepted by the app, validates them and push the associated article to `HomeViewController`

## Test Steps
* Build/run this branch
* Login if necessary
* Install one or more widgets
* Tap on a widget item
* Verify that it shows correctly, regardless where you were in the app
* Go back, and verify you are on Home

## PR Checklist:
- [x] Added Unit / UI tests
- [x] Self Review (review, clean up, documentation, run tests)
- [x] Basic Self QA

## Screenshots

Commit:
feat(universalLinks):

update MainViewModel, method handle(_ url: URL); fetch item from view context

Commit:
feat(universalLinks):

Add RouterTests

Commit:
feat(universalLinks):

Rename Router to UrlValidator

Gio2018 commented on PocketKit/Sources/PocketKit/Home/HomeViewController.swift:
```diff
@@ -136,6 +136,8 @@ class HomeViewController: UIViewController {
     override func viewDidLoad() {
         super.viewDidLoad()
         view.accessibilityIdentifier = "home"
+        navigationController?.delegate = self
```
> [!NOTE]
> this was missing in the existing implementation, thus the delegate methods, which have an implementation, were not called.

cyndichin replied:
ooh nice catch! I'm surprised we didn't notice any issue prior to this PR in terms of the delegate methods not being called.

cyndichin commented on PocketKit/Sources/PocketKit/Main/MainViewModel.swift:
```diff
@@ -217,12 +217,28 @@ class MainViewModel: ObservableObject {
+            return
+        }
+        // show the item associated to the given URL
+        if let savedItem = item.savedItem {
+            home.select(savedItem: savedItem, readableSource: .widget)
+        } else if let recommendation = item.recommendation {
+            home.select(recommendation: recommendation, readableSource: .widget)
+        }
```
do we want to handle the case if an `item` has no savedItem of recommendation associated? send a log to sentry? (tho the case shouldn't be trigger if its coming from the widgets)

Gio2018 replied:
> do we want to handle the case if an item has no savedItem of recommendation associated?

They would not come from a widget, so this will never happen here

cyndichin reviewed (commented):
great work @Gio2018 ! I ran through the test steps and I was brought to Home from Saves, Settings. But when I navigated to an item, collection or slate detail, I remained on that page instead of Home tab. Is that expected?

dskuza reviewed (commented):
Cyndi leaves a good (optional, imo) comment.

A few things: 

1. A "good" refactor (that probably shouldn't be addressed by this PR) would be to update some functions to not pass around `NSManagedObject`s as arguments to functions, but rather their object IDs.
2. Can we verify with Analytics that it's okay to not add an index to the analytics? I understand that it's optional per the data structure (and thus would be good in Snowplow Micro, I think?), but I'm wondering if Analytics, for some metrics, are allowing UI entities without an index. Maybe it's worth trying to grab the index of the item we're opening within the list, rather than sending nil. Not sure of the complexity of that, though.

CMasterson commented on PocketKit/Tests/PocketKitTests/UrlValidatorTests.swift:
```diff
@@ -0,0 +1,36 @@
+
+    func testWrongUrlReturnsNil() {
+        let validator = UrlValidator()
+        let index = Int.random(in: 0..<Self.validUrlNotFromWidgets.count)
```
Does this pull a random URL each time?

Gio2018 replied:
yes

CMasterson replied:
I really feel that tests should be the exact same every time they are run.
In this test case we could run through all the urls and make sure they all work as expected.

Gio2018 replied:
Fine: I'll just use one url then

CMasterson replied:
More is more better!

CMasterson commented on PocketKit/Tests/PocketKitTests/UrlValidatorTests.swift:
```diff
@@ -0,0 +1,36 @@
+    func testWrongUrlReturnsNil() {
+        let validator = UrlValidator()
+        let index = Int.random(in: 0..<Self.validUrlNotFromWidgets.count)
+        let wrongtUrlString = Self.validUrlNotFromWidgets[index]
```
What does 'wrong' mean in this instance?

Also: wrong((t))

Gio2018 replied:
good catch: probably best to rename as `invalid`

Gio2018 commented:
> 2. Can we verify with Analytics that it's okay to not add an index to the analytics? I understand that it's optional per the data structure (and thus would be good in Snowplow Micro, I think?), but I'm wondering if Analytics, for some metrics, are allowing UI entities without an index. Maybe it's worth trying to grab the index of the item we're opening within the list, rather than sending nil. Not sure of the complexity of that, though.

I looked at the analytics spreadsheet, and there were no specs about the UI element index (which, indeed, is optional). I can sure ask for clarification but won't hold this PR until after the all hands, in case the analytics team is not available (as it might be at this point Friday), would you agree @dskuza ?

CMasterson commented on PocketKit/Sources/Analytics/AppEvents/Home.swift:
```diff
@@ -92,7 +92,7 @@ public extension Events.Home {
      Fired when a user clicks a card on Home using the /discover API
      */
-    static func SlateArticleContentOpen(url: String, positionInList: Int, slateId: String, slateRequestId: String, slateExperimentId: String, slateIndex: Int, slateLineupId: String, slateLineupRequestId: String, slateLineupExperimentId: String, recommendationId: String, destination: ContentOpen.Destination) -> ContentOpen {
+    static func SlateArticleContentOpen(url: String, positionInList: Int?, slateId: String, slateRequestId: String, slateExperimentId: String, slateIndex: Int?, slateLineupId: String, slateLineupRequestId: String, slateLineupExperimentId: String, recommendationId: String, destination: ContentOpen.Destination) -> ContentOpen {
```
Maybe a bit late to the party, can we make a model to pass in here as one item instead of the many we have now?

Gio2018 replied:
this is legacy code, we can refactor but in a separate PR.

Gio2018 commented:
> great work @Gio2018 ! I ran through the test steps and I was brought to Home from Saves, Settings. But when I navigated to an item, collection or slate detail, I remained on that page instead of Home tab. Is that expected?

@cyndichin , would you mind showing me the repro steps in a live chat? I can't repro..

Commit:
feat(universalLinks):

update UrlValidatorTests, minor refactor

Commit:
feat(universalLinks):

update HomeViewController, remove UINavigationControllerDelegate methods as they are not used

CMasterson commented on PocketKit/Tests/PocketKitTests/UrlValidatorTests.swift:
```diff
@@ -0,0 +1,27 @@
+        XCTAssertNotNil(invalidUrl)
+        let itemUrl = validator.getItemUrl(from: invalidUrl!)
+        XCTAssertNil(itemUrl)
+    }
```
Can you add one more test that throws some random garbage URL into the validator?

Gio2018 replied:
That would be a test for `URL.init(_:)`, so I don't think it fits here.

CMasterson replied:
More what I was getting at was adding a valid URL that looks nothing like what the validator would be expecting

```let weirdoURL = URL(string: "testing:some.url.that.is.still.a/~valid/url?but=also.has&aBunch=of,stuff&array=that&array=we&array=dont&usually=have1234567890$-_.+!*'()")```

Gio2018 commented:
@dskuza , as discussed on Slack, I removed the `UINavigationControllerDelegate` methods from `HomeViewController`, since they were actually not needed.

Commit:
feat(universalLinks):

fix lint issue

cyndichin reviewed (approved):
tested again and was able to see the items now. thank you!
